src/finufft.cpp

```diff
@@ -89,6 +89,29 @@ namespace common {
 // Technically global state...
 // Needs to be static to avoid name collision with SINGLE/DOUBLE
 static std::mutex fftw_lock;
```

Comment:
It happens with one compiler, and only with optimizations on, with only one thread. We also weren't doing mixed precision planning in that case. It was failing on either the `double` _or_ `float` `dumbinputs` test. I confirmed that the lock was not being unset when it was supposed to during `FINUFFT_MAKEPLAN` routines (as best I could with my limited mac setup). This lock is implemented almost identically to `std::lock_guard` as well, so I'm not doing anything non-standard here. The only difference was using a global outside the class rather than a class variable, which is why I added the reference to the `static std::mutex` variable to the class private.

I plan on making a minimum working example independent of `finufft` for the `llvm` folk later to fully verify. As it stands I'm only _mostly_ sure that it's a clang issue, and not something more complicated in the language spec that I'm missing.

Reply:
can you run the example with sanitize undefined? This is something that usually happens when there is a undefined behavior slipping in.

Reply:
sanitizer is coming back clean

Reply:
I agree the most likely explanation is an issue with lifetimes in llvm.

Reply:
> This is inside a precision dependent source that is compiled into two separate objects, this means that there are two locks existing one for single precision one for double if the user uses single precision and double precision makeplan on multiple threads at the same time will this crash?

Wompwopm. I was mistaken here (verified with objdump and rereading docs). They are distinct copies of the same variable (same name, different address). I'll refactor to make this a proper global shared between compilation units. Might be better for another PR though.

Reply:
I was about to type: 
```bash
readelf -Ws libfinufft.a  | grep fftw_lock
    20: 0000000000000020    40 OBJECT  LOCAL  DEFAULT    6 _ZN7finufft6commonL9fftw_lockE
    20: 0000000000000020    40 OBJECT  LOCAL  DEFAULT    6 _ZN7finufft6commonL9fftw_lockE
```